### PR TITLE
Improved reassembly of fragments

### PR DESCRIPTION
**Pull request by mitza-oci**
* updated strategy for tracking completed fragments, now using DisjointSequence
* added expiration for incomplete fragments

### CODE REVIEW

File: dds/DCPS/transport/framework/TransportReassembly.cpp [hunk 1]
```diff
@@ -312,9 +313,11 @@ TransportReassembly::reassemble_i(const SequenceRange& seqRange,
     swap(data, iter->second.range_list_.front().rec_ds_);
     fragments_.erase(iter);
     completed_[key.publication_].insert(key.data_sample_seq_);
-    VDBG((LM_DEBUG, "(%P|%t) DBG:   TransportReassembly::reassemble_i "
-          "removed frag, returning %C with %d fragments\n",
-          data.sample_ ? "true" : "false", int(fragments_.size())));
+    if (Transport_debug_level > 5 || transport_debug.log_fragment_storage) {
+      ACE_DEBUG((LM_DEBUG, "(%P|%t) DBG:   TransportReassembly::reassemble_i "
+                 "removed frag, returning %C with %B fragments\n",
+                 data.sample_ ? "true" : "false", fragments_.size()));
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.cpp:319]:
```suggestion
                 data.sample_ ? "true (complete)" : "false (incomplete)", fragments_.size()));
```

File: dds/DCPS/transport/framework/TransportReassembly.cpp [hunk 2]
```diff
@@ -261,27 +266,22 @@ TransportReassembly::reassemble_i(const SequenceRange& seqRange,
...
     expiration_queue_.insert(std::make_pair(expiration, key));
     // since this is the first fragment we've seen, it can't possibly be done
-    VDBG((LM_DEBUG, "(%P|%t) DBG:   TransportReassembly::reassemble_i "
-          "stored first frag, returning false (incomplete) with %d fragments\n",
-          int(fragments_.size())));
+    if (Transport_debug_level > 5 || transport_debug.log_fragment_storage) {
+      ACE_DEBUG((LM_DEBUG, "(%P|%t) DBG:   TransportReassembly::reassemble_i "
+                 "stored first frag, returning false (incomplete) with %B fragments\n",
+                 fragments_.size()));
+    }
```
Comment (simpsont-oci) [inline dds/DCPS/transport/framework/TransportReassembly.cpp:284]: I can't put a comment there, but is it worth adding a log message for the completed case (line 290)?

**Review by simpsont-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/transport/framework/TransportReassembly.cpp (hunk 2 above)
Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.cpp:284]: It wasn't logged since it doesn't change `fragments_`.  Not that it couldn't be...

**Review by simpsont-oci — APPROVED** [review]

File: dds/DCPS/transport/framework/TransportReassembly.h [hunk 3]
```diff
@@ -96,19 +110,29 @@ class OpenDDS_Dcps_Export TransportReassembly {
...
   };
 
   typedef OPENDDS_MAP(FragKey, FragInfo) FragInfoMap;
   FragInfoMap fragments_;
 
+  typedef OPENDDS_MULTIMAP(MonotonicTimePoint, FragKey) ExpirationQueue;
+  ExpirationQueue expiration_queue_;
+
+  typedef OPENDDS_MAP_CMP(PublicationId, DisjointSequence, GUID_tKeyLessThan) CompletedMap;
+  CompletedMap completed_;
```
Comment (jrw972) [inline dds/DCPS/transport/framework/TransportReassembly.h:130]: Do we need to worry about `completed_` growing without bound?

Comment (mitza-oci) [reply dds/DCPS/transport/framework/TransportReassembly.h:130]: The entries from `completed_` will be removed when no longer associated with that Writer.
The entries in the `DisjointSequence` inside the map will grow in the same way that tracking sequence numbers in the datalink does.

**Pull request merged**